### ISSUE TITLE
how to use Generic Transformer or Specialized Transformer during inference

### ISSUE THREAD

**Issue by xueyingliu** [microsoft/DeepSpeed#4371]
In your [inference blog](https://www.microsoft.com/en-us/research/blog/deepspeed-accelerating-large-scale-model-inference-and-training-via-system-optimizations-and-compression/),  you talk about Generic Transformer and Specialized Transformer, 

> Generic Transformer replaces individual PyTorch operators within Transformer such as LayerNorm, Softmax, and bias-add with highly optimized DeepSpeed versions created using deep fusion.
> Specialized Transformer takes deep fusion one step further by creating fused schedules that not only fuse micro-operators within a PyTorch macro-operator (such as Softmax), but also fuse multiple macro-operators (such as Softmax and LayerNorm together with transpose op, and even GeMM). The fusion structure of our specialized Transformer kernel is shown in Figure 1.

Is the code of "Generic Transformer and Specialized Transformer" mentioned above released in this repo? 
If yes, how to use them in inference?

**Comment by xueyingliu** [comment]
Hi, there is another question, in your inference [Tutorials](https://www.deepspeed.ai/tutorials/)

> To inject the high-performance kernels, you need to set the replace_with_kernel_inject to True for the compatible models

Is 'the high-performance kernels' mean 'inference-optimized CUDA kernels' that include 'Deep fusion' and 'Inference-customized GeMM'?

**Comment by mrwyattii** [comment]
Hi @xueyingliu, the kernels referenced in this tutorial can be enabled with `replace_with_kernel_inject=True` when using DeepSpeed-Inference. For example if you want to utilize our custom kernels to accelerate inference on a Llama model, the following code could be used:
```python
# llama-ds-examply.py
import os
import torch
import deepspeed
from transformers import pipeline

local_rank = int(os.getenv("LOCAL_RANK", "0"))
world_size = int(os.getenv("WORLD_SIZE", "1"))
prompt = "DeepSpeed is"

pipe = pipeline(
    "text-generation",
    "huggyllama/llama-7b",
    torch_dtype=torch.float16,
    device=torch.device(f"cuda:{local_rank}"),
)

pipe.model = deepspeed.init_inference(
    model=pipe.model,
    mp_size=world_size,
    dtype=torch.float16,
    replace_method="auto",
    replace_with_kernel_inject=True,
)

response = pipe(prompt)
print(response)
```
Run with `deepspeed --num_gpus 1 llama-ds-example.py`

**Comment by xueyingliu** [comment]
@mrwyattii Thanks for your reply, and where is the source code?

**Comment by mrwyattii** [comment]
Source code can be found here for the inference kernels: https://github.com/microsoft/DeepSpeed/tree/master/csrc/transformer/inference

**Comment by xueyingliu** [comment]
@mrwyattii OK
